### PR TITLE
Concatenation optimization

### PR DESCRIPTION
**Pull request by vhavlena**
This PR introduces new optimizations concerning concatenation

- inplace concatenation
- new operations for `Delta` to speed-up particular automata algorithms

### CODE REVIEW

**Comment by vhavlena** [comment]
I need to resolve the conflicts somehow first

**Comment by vhavlena** [comment]
Ok, it could be ready for review now. There are still things to discuss but this in-place concatenation should be more efficient than the previous one.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:299]: I would add a note to the comment that it can break sortedness.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: The `IMPORTANT` note is not a part of the `brief` section. 
```suggestion
     * targets.
     *
     * IMPORTANT: In order to work properly, the lambda function needs to be 
```

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: This is exactly [`std::transform()`](https://en.cppreference.com/w/cpp/algorithm/transform), maybe we could keep the name `transform()`, seeing that it is idiomatic for this operation in many programming languages.

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: Why not just overload `add()`?

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh]: We agreed that methods should operate in-place by default. Maybe call this just `concatenate()` (with optional chaining support).

The function `concatenate()` returning a new NFA would be then the out-of-place concatenation accepting two automata as parameters: `const Nfa& a, const Nfa& b`.

File: include/mata/ord-vector.hh
Comment (Adda0) [inline include/mata/ord-vector.hh]: I would say that we are trying to keep the opening bracket on the same line as the function definition. I would revert this change.
```suggestion
    void insert(iterator itr, const Key& x) {
```

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Use [`emplace_back()`](https://en.cppreference.com/w/cpp/container/vector/emplace_back) for the Move and call `std::move()` for the other `push_back()` calls. It should be more performant.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Would it be more efficient to copy the whole original delta first and only then traverse the target states to apply lambda? Or simply call `std::transform()` in the body of this function?

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Use a more descriptive function.
```suggestion
    if(states.empty()) {
```

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Use `const` for possible optimizations from the compiler.
```suggestion
void Delta::add_set(const State state_from, const Symbol symbol, const StateSet& states) {
```

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Shouldn't it be `max_state + 1` for the parameter `needed_capacity`?

File: include/mata/nfa.hh
Comment (Adda0) [inline include/mata/nfa.hh:1]: We should write Catch2 tests for each of the newly introduced functions.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: If you rebase on the newest `devel`, you should be able to just look for `symbol` instead of having to create the `Move` object.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Fix both comments to use the new variable names.

File: src/nfa/nfa.cc
Comment (Adda0) [inline src/nfa/nfa.cc]: Do not use `const` here. It will prevent the compiler to do a move operation. Create the object as an R-value reference inside the `insert()` method call as a temporary object.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [inline src/nfa/nfa-concatenation.cc]: Do not create the variable, pass the result of the call to `copy_posts_with()` to `append()` directly. We should also think about implementing `append()` with parameter passed by value and using `std::move` here instead of passing by reference.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [inline src/nfa/nfa-concatenation.cc]: We need to rebase onto `devel` and use `SparseSet` here instead.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [inline src/nfa/nfa-concatenation.cc]: Unused variable.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [inline src/nfa/nfa-concatenation.cc:55]: Do we try to resolve these TODOs in this PR?

**Review by Adda0 — COMMENTED** [review]
Looks good to me. This should be a great improvement.

I would create another overload, appending the whole delta (it would do the same as both `append()` and `insert()` operations). It would only call functions `append()` and `insert()` with its `posts` member variable as a parameter.

If we later change the underlying structure of `Delta`, we would only have to change the implementation in this overloaded function, not all the uses of `std::vector<Post>` as well. This is useful also because as of now, `Delta` does not have a method to get all posts for easy calling `append()` with a vector of posts as a parameter.

I also include a [patch](https://gist.github.com/Adda0/c8ae75c15fdeddd2d09a051e32c10ee6) to apply containing little optimizations and improvements. See for yourself.

File: include/mata/nfa.hh
Comment (vhavlena) [reply include/mata/nfa.hh:299]: In fact it cannot break sortedness. However, it may have an unexpected behaviour.

Comment (Adda0) [reply include/mata/nfa.hh:299]: My mnistake. The index of first post in the appended `post_vector` will be changed to index `size()` in delta. I agree that if this is the intended behaviour, it cannot break sortedness. But it is not the same as merging two deltas together. We should therefore clarify this different. I already assumed wrongly about these indices after the append is performed.

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc]: I don't know, but I think there will be no real difference (concerning the performance).

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc]: Yes

File: include/mata/nfa.hh
Comment (vhavlena) [reply include/mata/nfa.hh:1]: Yes

File: src/nfa/nfa-concatenation.cc
Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc:55]: I suggest to postpone. It might require some discussion because you want to apply lambda on each element before inserting.

File: include/mata/nfa.hh
Comment (vhavlena) [reply include/mata/nfa.hh:299]: It is a different thing imo. The method do not claim that it appends two Deltas.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [reply src/nfa/nfa-concatenation.cc:55]: Sounds good to me. Closing.

File: include/mata/nfa.hh
Comment (Adda0) [reply include/mata/nfa.hh:299]: Yeah, it is. I just confused the two together because we talked earlier about having a function to merge two deltas. OK, you convinced me that it is not necessary to explain here any further.

**Comment by Adda0** [comment]
And of course, we need to eliminate any and all warnings before merging the PR.

**Comment by vhavlena** [comment]
@jurajsic and @kilohsakul could you take a look on the changes? I would like to merge it and see if it has some impact on regex-intense benchmarks in z3-noodler.

**Comment by Adda0** [comment]
In the meantime, could you rebase onto the current `devel` and if there are any warnings, try to fix them? I think there are none from looking at the compilation logs, but just to be sure. If you do not want to work on this, tell me and I will do the rebase.

I will approve the PR when everyone gave their reviews, and we have resolved all discussed topics.

File: src/nfa/nfa-concatenation.cc
Comment (jurajsic) [inline src/nfa/nfa-concatenation.cc]: Why do you not take `this->delta[upd_fnc(ini)]` here? You won't have to apply lambda when inserting later. And we could create the function for adding Post to other Post already in this PR then.

**Review by jurajsic — APPROVED** [review]

File: src/nfa/nfa-concatenation.cc
Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc]: 1) good point, I change it
2) there is `add` method of `Delta` for adding a set of states. The problem is that according to my shallow experiments it works worse than adding it one by one (as it is now)

**Comment by vhavlena** [comment]
> In the meantime, could you rebase onto the current `devel` and if there are any warnings, try to fix them? I think there are none from looking at the compilation logs, but just to be sure. If you do not want to work on this, tell me and I will do the rebase.
> 
> I will approve the PR when everyone gave their reviews, and we have resolved all discussed topics.

Can you do it please? I think it should be ok as the warnings I am getting are not related to this PR

File: src/nfa/nfa-concatenation.cc
Comment (jurajsic) [reply src/nfa/nfa-concatenation.cc]: Which method?

Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc]: Override of `Delta::add` taking the third argument a set of states.

Comment (jurajsic) [reply src/nfa/nfa-concatenation.cc]: Isn't that your method? Also I just realized, can you move it to the other two `add` methods so they are next to each other?
But what I meant is that maybe Post should have a method that adds another Post to it (@kilohsakul is already telling me his genius idea how to do it).

**Review by tfiedor — APPROVED** [review]
Beautiful code. No issues. All hath been said by other fellows.

**Comment by Adda0** [comment]
> Can you do it please? I think it should be ok as the warnings I am getting are not related to this PR

I am on it.

**Comment by Adda0** [comment]
By the way, do not ever add profiling tests to the normal Catch2 tests framework without turning them off by default, using a tag beginning with `.` at the start of the tag name. We usually use `⁣.profiling` tag: `TEST_CASE("Concat_inplace performance", "[.profiling]") {`. My notebook nearly burnt down before I figured out what is going on ;)

File: src/nfa/nfa-concatenation.cc
Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc]: Yes, I added it. I understand, but I need to think about it. Currently I don't see why Post::add should be significantly better than adding it one by one

**Comment by vhavlena** [comment]
> By the way, do not ever add profiling tests to the normal Catch2 tests framework without turning them off by default, using a tag beginning with `.` at the start of the tag name. We usually use `⁣.profiling` tag: `TEST_CASE("Concat_inplace performance", "[.profiling]") {`. My notebook nearly burnt down before I figured out what is going on ;)

Strange. On my laptop it is not somehow demanding (approx 2 sec)

**Comment by Adda0** [comment]
Then there is a bug somewhere. It starts allocating a lot of memory ~~indefinitely~~ on my machine. Up to 500 concatenation iterations, it finishes in reasonable time (a few seconds), but the time increases and the allocated space goes up to .5 GB and counting. I will investigate tomorrow.

For now, it seems that it is just demanding on my machine. But it works correctly, and there is no bug, probably.

**Comment by tfiedor** [comment]
> Then there is a bug somewhere. It starts allocating a lot of memory ~indefinitely~ on my machine. Up to 500 concatenation iterations, it finishes in reasonable time (a few seconds), but the time increases and the allocated space goes up to .5 GB and counting. I will investigate tomorrow.
> 
> For now, it seems that it is just demanding on my machine. But it works correctly, and there is no bug, probably.

How are you running the PR? `make release` + `make test`? I could try it in my environment. It might help you with debug (my guess is you have something set in your IDE, or you compile in debug mode---still this might show there is some issue in the code).

**Comment by Adda0** [comment]
Yeah, I am running all tests in debug mode unless I am doing performance testing. That is what I suspect is the whole problem, but I did not investigate now. I will have a look to satisfy our curiosity… And yes, if I compile in release mode, I get 1.8 s runtime. Therefore, I would close the problem as resolved.

**Review by Adda0 — APPROVED** [review]
There are no warnings and the PR looks good. I approve merging the PR.

File: include/mata/ord-vector.hh
Comment (kilohsakul) [inline include/mata/ord-vector.hh:202]: Are you sure the current merge is slow?
I have an idea for an inplace merge, using a trick that swaps elements between the two vectors, but is it worth it?

File: include/mata/nfa.hh
Comment (kilohsakul) [inline include/mata/nfa.hh:562]: Does it destroy also the automaton in the argument? Add this (whether it destroys the argument or no) to the comment?

File: src/nfa/nfa-concatenation.cc
Comment (kilohsakul) [inline src/nfa/nfa-concatenation.cc:42]: couldn't you use this? If not, maybe it should be fixed in the SparseSet?
Util::SparseSet<State> new_fin{aut.final};

Comment (kilohsakul) [inline src/nfa/nfa-concatenation.cc:42]: should we do it using SparseSet.rename()?
Currently not sure that it would work, since it was made assuming that renaming gives a smaller name, where here it gives a larger name. Myabe we would need two variants or to fix it somehowe, but anyway, should we do such things using methods like rename?

File: src/nfa/nfa-concatenation.cc
Comment (kilohsakul) [inline src/nfa/nfa-concatenation.cc:48]: isn't is too much optimization? Line 48 is omre expensive but constant too. Just wandering

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:333]: Is this already the best we can do or we ultimately need something better that avoids searching, like the same thing but parameterized with two iterators instead of state and symbol?

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:344]: hm, these atransitions are not transitions as (q,a,r), right?
Would be good to invent some terminology for these things, everybody uses his own language.

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:349]: The constructor of Move should probably be made explicity so that we have to write Move{...

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:508]: do we want to rename num_of_states, so that we don't have to search for definition?

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:518]: is Move a good name? :)

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:522]: I am actually giving up, probably need more comments.

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:2296]: Is this comment at the right place?

File: src/nfa/nfa-concatenation.cc
Comment (kilohsakul) [reply src/nfa/nfa-concatenation.cc:55]: but lets try to remember this, to eventually figure out how to do these things systematically and easily

File: src/nfa/nfa-concatenation.cc
Comment (kilohsakul) [inline src/nfa/nfa-concatenation.cc:52]: A comment would be good, 
maybe also bacause it is a bit weird, it adds the same final states multiple times if there are more states that are both initial and final

File: include/mata/nfa.hh
Comment (Adda0) [reply include/mata/nfa.hh:562]: It cannot destroy the second automaton: the argument is passed as a const reference.

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [reply src/nfa/nfa-concatenation.cc:48]: If you check something in a for loop multiple times, it is always worth extracting the test outside the for loop, in my opinion. Even if the optimizer will most probably do the same thing as we do. We do the same with the condition check in the for loop itself. If we check for, for example, `i < delta.size()`, we rewrite it as `const size_t delta_size = delta.size(); for (size_t i{ 0 }; i < delta_size; ++i)`.

File: src/nfa/nfa.cc
Comment (Adda0) [reply src/nfa/nfa.cc:333]: We can add an overloaded function with two iterators. We will see whether we find usages for it in our algorithms.

File: src/nfa/nfa.cc
Comment (Adda0) [reply src/nfa/nfa.cc:344]: It is a `Post`. We can just call it `post`, which means exactly transitions from a single state. That is, state transitions. We just need to get used to using `post` and `move` as descriptive names for variables.

File: src/nfa/nfa.cc
Comment (Adda0) [reply src/nfa/nfa.cc:508]: What do you propose? `size()` is already taken. This is exactly what the value represents. If we name it something else, we will have to search for definition to see that it represents the number of states in delta.

File: include/mata/nfa.hh
Comment (kilohsakul) [reply include/mata/nfa.hh:562]: okok

File: include/mata/ord-vector.hh
Comment (vhavlena) [reply include/mata/ord-vector.hh:202]: For concatenation it makes just a small sense to me (we are adding items to the back, which seems to be fast now). But I guess there are some operations using this Union, so maybe we could take a look.

File: src/nfa/nfa-concatenation.cc
Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc:42]: Probably yes, in general. In this simple case I am not sure. But I am pretty sure that it won't be more efficient (you need to copy the original SparseSet and then to rename). Here it is a single loop only.

File: src/nfa/nfa-concatenation.cc
Comment (vhavlena) [reply src/nfa/nfa-concatenation.cc:52]: Yes it is simpler than to track already added states (which is exactly what SparseSet::insert does)

File: src/nfa/nfa-concatenation.cc
Comment (kilohsakul) [reply src/nfa/nfa-concatenation.cc:48]: yes yes, but I am always wandering whether these are needles eager optimizations that make things uglier and hard to read

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:344]: Get used to it or invent better names, the current move and post are maybe survivable but feel arbitrary. It is hard though, requires spending brain cycles. If someone has some idea, throuw it here.

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:508]: something as num_of_incident/touched_states. Besides, num is somewhat missleading, max would be better if 0 was not ambiguous ... ?

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:333]: That was the question, maybe it is for vojta: could an iterator version save some searching?

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:516]: isn't "lambda" like calling a variable "variable"? :)

File: src/nfa/nfa.cc
Comment (kilohsakul) [inline src/nfa/nfa.cc:506]: Is there a reason for why this is not in place? Or should it be?

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:2296]: I think so. It is just a multiline comment related the following code.

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:506]: Yes I have made it deliberately const. It avoids copying of Delta inside concatenation. The second thing is that after in-place modification Delta could by inconsistent.

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:516]: Or "i" for index? Come on, stop generating useless comments.

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:508]: I am indifferent. num_of_states doesn't offend me, but if you have a better idea, let's rename it

File: src/nfa/nfa.cc
Comment (vhavlena) [reply src/nfa/nfa.cc:333]: Rather no for the case of concatenation. But it might be useful later when we dive into optimization of other operations.

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:508]: just that we did guess it both wrong with juraj

File: src/nfa/nfa-concatenation.cc
Comment (Adda0) [reply src/nfa/nfa-concatenation.cc:55]: I have a note in my TODO list. Closing the conversation.

**Comment by vhavlena** [comment]
Can we merge it ?

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:506]: Ok, I get there is some reason, but I don't get what it is. Is it perhaps possible to write things in  even more idiot friendly way, at least when there is time?

File: src/nfa/nfa.cc
Comment (kilohsakul) [reply src/nfa/nfa.cc:2296]: Ok. It is confusing because it says insert without searching and the next line is searching.

File: src/nfa/nfa.cc
Comment (Adda0) [reply src/nfa/nfa.cc:506]: This is what was designed in C++ to create a simple interface for passing functions around. See [SO answer](https://stackoverflow.com/a/20353798) explaining what `std::function` is and how to use it.

The implemented function exactly simulates standard `std` function [`std::transform`](https://en.cppreference.com/w/cpp/algorithm/transform) or [`std::ranges::transform`](https://en.cppreference.com/w/cpp/algorithm/ranges/transform) which should give an idea of what the function does immediately.

**Comment by kilohsakul** [comment]
> Can we merge it ?

Yes let's merge it. No need to get stuck particularly on my comments.

Just, in general, if we could put more effort to making code easy to read, at least when we are not under preasure, that would be great. My feeling is that this one could  be made really easy to read, mainly by commenting (I will try to do this in things I wrote).

**Review by kilohsakul — APPROVED** [review]

**Pull request merged**
